service/src/main/java/org/jboss/sbomer/service/feature/sbom/rest/criteria/AbstractCriteriaAwareRepository.java

```diff
@@ -0,0 +1,141 @@
+
+    CriteriaBuilder criteriaBuilder;
+    EntityManagerAdapter entityManagerAdapter;
+    Class<T> entityType;
+
+    static {
+        Set<ComparisonOperator> predicateOperators = RSQLOperators.defaultOperators();
+        predicateOperators.add(IS_NULL);
+        predicateOperators.add(IS_EQUAL);
+
+        predicateParser = new RSQLParser(predicateOperators);
+
+        Set<ComparisonOperator> sortOperators = new HashSet<ComparisonOperator>();
+        sortOperators.add(ASC);
+        sortOperators.add(DESC);
+
+        sortParser = new RSQLParser(sortOperators);
+    }
+
+    /**
+     * Reads the total number of entities that satisfy the RSQL query.
+     *
+     * @param rsqlQuery RSQL query to be taken into account.
+     * @return
+     */
+    public Long countByRsqlQuery(String rsqlQuery) {
+        CriteriaQuery<Long> criteriaQuery = criteriaBuilder.createQuery(Long.class);
+        Root<T> root = criteriaQuery.from(entityType);
+        criteriaQuery.select(criteriaBuilder.countDistinct(root));
+
+        criteriaQuery = handleRsql(criteriaQuery, root, rsqlQuery);
+        return getEntityManager().createQuery(criteriaQuery).getSingleResult();
+    }
+
+    public AbstractCriteriaAwareRepository(Class<T> entityType) {
+        this.entityType = entityType;
+    }
+
+    @PostConstruct
+    @Transactional
+    protected void init() {
+        entityManagerAdapter = new EntityManagerAdapter(getEntityManager());
+        criteriaBuilder = entityManagerAdapter.getCriteriaBuilder();
+    }
+
+    protected <X> CriteriaQuery<X> handleRsql(CriteriaQuery<X> query, Root<T> root, String rsqlQuery) {
+
+        if (Strings.isEmpty(rsqlQuery)) {
+            return query;
+        }
+
+        log.debug("Applying provided RSQL query to enhance search: '{}'", rsqlQuery);
+
+        // Create custom implementation of RSQLVisitor which converts nodes to predicates using
+        // CustomizedPredicateBuilder or CustomizedPredicateBuilderStrategy for custom operators
+        RSQLVisitor<Predicate, EntityManagerAdapter> visitor = new CustomizedJpaPredicateVisitor<X>().withRoot(root)
+                .withPredicateBuilderStrategy(new CustomizedPredicateBuilderStrategy());
+
+        // create RSQLParser with default and custom operators
+        Node rootNode = predicateParser.parse(rsqlQuery);
+        Predicate predicate = rootNode.accept(visitor, entityManagerAdapter);
+
+        return query.where(predicate);
+    }
+
+    protected <X> CriteriaQuery<X> handleSort(CriteriaQuery<X> query, Root<T> root, String sort) {
+        if (Strings.isEmpty(sort)) {
+            return query;
+        }
+
+        String compliantSort = CustomPredicateSortBuilder.rsqlParserCompliantSort(sort);
+        log.debug("Modified RSQL sort string from: '{}' to a RSQL parser compliant format: '{}'", sort, compliantSort);
+
+        CustomizedJpaPredicateSortVisitor<?> sortVisitor = new CustomizedJpaPredicateSortVisitor<>(Sbom.class)
```

Comment:
This is probably not an `Sbom.class` but a more generic `entityType` ? So it can be used by all entity classes?

Reply:
Good catch, I think I was using it for testing. It should be here: `entityType` instead.